rasterio/rio/features.py

```diff
@@ -20,6 +19,128 @@
 logger = logging.getLogger('rio')
 
 
+# Extract command
+@cli.command(short_help='Extract raster using features.')
+@files_inout_arg
+@format_opt
+@click.option('--all_touched', is_flag=True, default=False,
+              help='Use all pixels touched by features for extraction, '
+              'otherwise use only pixels whose center is within the '
+              'polygon or that are selected by Brezenhams line algorithm')
+@click.option('--crop', is_flag=True, default=False,
+              help='Crop output raster to the extent of the geometries. '
+                   'GeoJSON must overlap input raster to use --crop')
+@click.option('--invert', is_flag=True, default=False,
+              help='Inverts the mask, so that areas covered by features are'
+                   'masked out and areas not covered are retained.  Ignored '
+                   'if using --crop')
+@click.pass_context
+def extract(
+        ctx,
+        files,
+        driver,
+        all_touched,
+        crop,
+        invert):
+
+    """Extracts pixels from all bands of a raster using features, masking out
+    all areas not covered by features, and optionally crops the output raster
+    to the extent of the features.  Features are assumed to be in the same
+    coordinate reference system as the input raster.
+
+    GeoJSON must be the first input file or provided from stdin:
+
+    > rio extract features.json input.tif output.tif
+
+    > rio extract input.tif output.tif < features.json
+
+    If the output raster exists, it will be completely overwritten with the
+    results of this operation.
+
+    The result is always equal to or within the bounds of the input raster.
+
+    --crop and --invert options are mutually exclusive.
+
+    --crop option is not valid if features are completely outside extent of
+    input raster.
+    """
```

Comment:
@sgillies Interesting idea, I'll have to think on it...

I do worry about use of ```-v``` here, since you are using it for verbosity [elsewhere](https://github.com/mapbox/rasterio/blob/master/rasterio/rio/info.py#L71), and ```-f``` is used here and in cligj for [output format](https://github.com/mapbox/cligj/blob/master/cligj/__init__.py#L38) 

I was trying to keep the usage pretty similar to ```rasterize``` (in terms of stdin or filename) so that it felt familiar to folks, and thinking there would be use cases where we pipe in GeoJSON rather than have it in advance, and this usage looks a little [awkward in grep](http://stackoverflow.com/questions/6641731/read-expression-for-grep-from-standard-input).

What do you expect the output of bounding box extraction to be, as specified above?   Original extent of input.tif, but masked out everywhere except [west, south, east, north]?  Or cropped to that extent as well?  When I was considering whether or not I should add this, it looks like gdal_translate covers the crop to extent of bbox pretty well, but doesn't combine that with masking (but result would be the same if we are only specifying a rectangular bbox in world coordinates of raster).